src/aof.c

```diff
@@ -1963,37 +1963,64 @@ static int rioWriteHashIteratorCursor(rio *r, hashTypeIterator *hi, int what) {
 /* Emit the commands needed to rebuild a hash object.
  * The function returns 0 on error, 1 on success. */
 int rewriteHashObject(rio *r, robj *key, robj *o) {
+    int res = 0; /*fail*/
+
     hashTypeIterator *hi;
     long long count = 0, items = hashTypeLength(o, 0);
 
+    /* Is expected also hash-fields with expiration (HFE) ? */
+    int isHFE = hashTypeGetMinExpire(o) != EB_EXPIRE_TIME_INVALID;
     hi = hashTypeInitIterator(o);
-    while (hashTypeNext(hi, 0) != C_ERR) {
-        if (count == 0) {
-            int cmd_items = (items > AOF_REWRITE_ITEMS_PER_CMD) ?
-                AOF_REWRITE_ITEMS_PER_CMD : items;
 
-            if (!rioWriteBulkCount(r,'*',2+cmd_items*2) ||
-                !rioWriteBulkString(r,"HMSET",5) ||
-                !rioWriteBulkObject(r,key)) 
-            {
-                hashTypeReleaseIterator(hi);
-                return 0;
+    if (!isHFE) {
+        while (hashTypeNext(hi, 0) != C_ERR) {
+            if (count == 0) {
+                int cmd_items = (items > AOF_REWRITE_ITEMS_PER_CMD) ?
+                                AOF_REWRITE_ITEMS_PER_CMD : items;
+                if (!rioWriteBulkCount(r, '*', 2 + cmd_items * 2) ||
+                    !rioWriteBulkString(r, "HMSET", 5) ||
+                    !rioWriteBulkObject(r, key))
+                    goto reHashEnd;
             }
+
+            if (!rioWriteHashIteratorCursor(r, hi, OBJ_HASH_KEY) ||
+                !rioWriteHashIteratorCursor(r, hi, OBJ_HASH_VALUE))
+                goto reHashEnd;
+
+            if (++count == AOF_REWRITE_ITEMS_PER_CMD) count = 0;
+            items--;
         }
+    } else {
+        while (hashTypeNext(hi, 0) != C_ERR) {
+            if (hi->expire_time == EB_EXPIRE_TIME_INVALID)
+            {
+                if (!rioWriteBulkCount(r, '*', 4) ||
+                    !rioWriteBulkString(r, "HMSET", 5) ||
+                    !rioWriteBulkObject(r, key))
+                    goto reHashEnd;
+            } else {
+                /* HSETF key PXAT msec FVS 1 field value */
+                char cmd[] = "*8\r\n$5\r\nHSETF\r\n";
```

Comment:
the reason why we don't rewrite it as single command(max AOF_REWRITE_ITEMS_PER_CMD item in one command) is that we don't support it now?
it seems werid that we will generate three command if there are three fields with TTL after aof rewrite.

Reply:
i still worried about wether this code will cause some performance problems, if we don't handle it in this PR, please add a note in the top comment to avoid forgetting it.